include/etl/iterator.h

```diff
@@ -894,7 +894,6 @@ namespace etl
   }
 #endif
 
-#if ETL_NOT_USING_STL || ETL_CPP17_NOT_SUPPORTED
```

Comment:
You cannot remove this conditional statement as std::size and etl::size cannot exist concurrently.
This is due to that fact that any use of ADL with this function will raise an 'ambiguous function error when using the STL.

e.g.
using std::size;
size_t s = size(container);

Reply:
The `using` is necessary due to the way overloads of STL template functions have to be declared.
As you are not allowed to define them in the `std` namespace, you usually define them in the global namespace instead.

When they are used in a template function, the compiler doesn't know whether the STL version or a custom user overload will be the best fit for the type.
Therefore you must use the `using`.

All of the following STL template functions have this issue...
```
swap()
begin()
end()
size()
```

Therefore, in any template code you must do something like the following...

```
template <typename TObject>
void DoSomthing(const TObject obj)
{
  using std::size; // Allow the automatic use of std::size.

  size_t s = size(obj); // ADL lookup. May use std::size or a custom overload.

  // Etc...
}
```

If you want code that can be configured to use either the ETL or STL then you must do the following.

```
template <typename TObject>
void DoSomthing(const TObject obj)
{
  using ETL_OR_STD::size;

  size_t s = size(obj); // ADL lookup. May use std::size, etl::size (depending on configuration) or a custom overload.

  // Etc...
}
```
_Do you have any alternative suggestions?_
What issue were you trying to solve by removing the conditional compilation?

The ETL defines its version of `size()` if either you are not using the STL or you _are_ using it but you are not using C++17.
C++17 was the first version that defined `std::size()`.

Reply:
The tests are using `std::size()` in many places. Because as you stated, C++17 is the first one that had `std::size()`, they won't compile when building the tests for C++14. `etl::size()` needs to be used.

Reply:
I did not know of the `ETL_OR_STD`. I can replace all uses of `std/etl::size` with that.

Reply:
I just tried `ETL_OR_STD` and it doesn't work with cpp14, because there isn't any `size` declared in `std` in cpp14. My first thought was to create a `size_for_etl_tests.h` with the following contents:

```c++
namespace etl_tests{
#if ETL_CPP17_NOT_SUPPORTED || ETL_NOT_USING_STL
using etl::size;
#else
using std::size;
#endif
};
```
Then use `etl_tests::size()` everywhere `std::size` is needed.

Reply:
I do have a question. It would seem that the only time a collision would occur is if `using std::size;` was declared inside the etl namespace. If `etl::size` is already defined, and works exactly like the one defined in `std`, why would you use the one outside the etl library?

Reply:
Getting the ETL to work seamlessly both with and without the STL has been a headache at times!

Reply:
Check the definitions in `etl/private/choose_namespace.h`

The are currently definitions for `ETL_OR_STD` and `ETL_OR_STD11` 
Maybe add a `ETL_OR_STD17`?

Reply:
Yeah, I noticed that. I already copied the STD11 one and created a STD17 one. I did a push and all the tests appear to still be passing.